rover_video/src/video_recording.cpp

```diff
@@ -0,0 +1,257 @@
+ * @return Recording&
+ */
+Recording& Recording::operator=(Recording&& other) noexcept
+{  // move operator just to be safe
+    if (this != &other)
+    {  // Prevent self-assignment
+
+        // Move resources
+        _RequestShutdown = std::move(other._RequestShutdown);
+        _recordingThread = std::move(other._recordingThread);
+        _stopRecording.store(other._stopRecording.load(std::memory_order_acquire), std::memory_order_release);
+
+        _camURL = std::move(other._camURL);
+        _pipeline = std::move(other._pipeline);
+        _filename = std::move(other._filename);
+        _videoFolderPath = std::move(other._videoFolderPath);
+        _files = std::move(other._files);
+        _recordingNumber = other._recordingNumber;
+        _startTime = other._startTime;
+        _frame_width = other._frame_width;
+        _frame_height = other._frame_height;
+        _fps = other._fps;
+        rLogger = std::move(other.rLogger);
+
+        _appender = std::move(other._appender);
+        _cap = std::move(other._cap);  // Move cv ressources
+        _video_writer = std::move(other._video_writer);
+        _frame = std::move(other._frame);
+    }
+    return *this;
+}
+
+/**
+ * @brief Destroy the Recording:: Recording object \n
+ * @brief Handle the release of CV objects
+ *
+ */
+Recording::~Recording(void)
+{
+    if (_cap.isOpened())  // avoid unnecessary logging when creating temporary objects
+    {
+        _stopRecording.store(true);
+
+        if (_recordingThread.joinable())
+        {
+            _recordingThread.join();
+        }
+        // Release resources
+        _cap.release();
+        _video_writer.release();
+        _appender.release();
+
+        RCLCPP_INFO(*rLogger, "Recording stopped.");
+    }
+}
+
+/**
+ * @brief initialize all CV variables
+ *
+ * @return true if all CV variables are initialized correctly \n
+ * @return false if there's any error
+ */
+bool Recording::startRecording(void)
+{
+    // Getting the directory for the recording
+    std::string filePath = _videoFolderPath + "/" + _filename;
+    filePath.insert(filePath.length() - 4, '_' + std::to_string(_recordingNumber++));  // add recording number before .avi
+    _files.push_back(filePath);                                                        // add file to list of recordings
+
+    _pipeline = "rtspsrc location=" + _camURL
+                + " latency=0 drop=true ! decodebin ! videorate max-rate=30 ! videoconvert ! queue max-size-buffers=1 ! appsink";
+
+    _cap.open(_pipeline, cv::CAP_GSTREAMER);
+    if (!_cap.isOpened())
+    {
+        RCLCPP_ERROR(*rLogger, "Failed to open camera stream.");
+        return false;
+    }
+
+    // Get frame width and height
+
+    _frame_width = static_cast<int>(_cap.get(cv::CAP_PROP_FRAME_WIDTH));
+    _frame_height = static_cast<int>(_cap.get(cv::CAP_PROP_FRAME_HEIGHT));
+    _fps = static_cast<double>(_cap.get(cv::CAP_PROP_FPS));
+
+    _fps = (_fps > 0) ? _fps : 30;
```

Comment:
You can use our CONSTRAIN macro for this, also FPS is double so using int is undefined

```suggestion
    _fps = CONSTRAIN(_fps, 0.0F, 30.0F);
```